File: images/quicksort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.BiPredicate;

public class QuickSort {

    public static <T> List<T> quicksort(List<T> arr, BiPredicate<T, T> less) {
        if (arr.size() <= 1) {
            return arr;
        }
        T pivot = arr.get(arr.size() / 2);
        List<T> left = new ArrayList<>();
        List<T> middle = new ArrayList<>();
        List<T> right = new ArrayList<>();
        for (T item : arr) {
            if (less.test(item, pivot)) left.add(item);
            else if (!less.test(item, pivot) && !less.test(pivot, item)) middle.add(item);
            else right.add(item);
        }
        List<T> sorted = new ArrayList<>();
        sorted.addAll(quicksort(left, less));
        sorted.addAll(middle);
        sorted.addAll(quicksort(right, less));
        return sorted;
    }

    public static void main(String[] args) {
        System.out.println(quicksort(Arrays.asList(3,6,8,10,1,2,1), (Integer a, Integer b) -> a < b));
        System.out.println(quicksort(Arrays.asList("apple", "grape", "banana", "cherry"), (String a, String b) -> a.compareTo(b) < 0));
    }
}

File: images/quicksort.rs
fn quicksort<T: Ord + Clone>(arr: &[T]) -> Vec<T> {
    if arr.len() <= 1 {
        return arr.to_vec();
    }
    let pivot = &arr[arr.len() / 2];
    let left: Vec<T> = arr.iter().filter(|&&x| x < *pivot).cloned().collect();
    let middle: Vec<T> = arr.iter().filter(|&&x| x == *pivot).cloned().collect();
    let right: Vec<T> = arr.iter().filter(|&&x| x > *pivot).cloned().collect();

    let mut result = quicksort(&left);
    result.extend(&middle);
    result.extend(&quicksort(&right));
    return result;
}

fn main() {
    let numbers = vec![3,6,8,10,1,2,1];
    let sorted_numbers = quicksort(&numbers);
    println!("{:?}", sorted_numbers);

    let words = vec!["apple", "grape", "banana", "cherry"];
    let sorted_words = quicksort(&words);
    println!("{:?}", sorted_words);
}
